src/herder/Upgrades.cpp

```diff
@@ -318,4 +329,348 @@ Upgrades::timeForUpgrade(uint64_t time) const
+            .first;
+    if (asset.type() != ASSET_TYPE_NATIVE && accountID == getIssuer(asset))
+    {
+        return;
+    }
+    if (iter->second)
+    {
+        if (!stellar::addBalance(*iter->second, delta))
+        {
+            iter->second.reset();
+        }
+    }
+}
+
+static int64_t
+getAvailableBalance(AccountID const& accountID, Asset const& asset,
+                    int64_t balanceAboveReserve, LedgerDelta& ld, Database& db)
+{
+    if (asset.type() == ASSET_TYPE_NATIVE)
+    {
+        return balanceAboveReserve;
+    }
+
+    if (accountID == getIssuer(asset))
+    {
+        return INT64_MAX;
+    }
+    else
+    {
+        auto trust = TrustFrame::loadTrustLine(accountID, asset, db, &ld);
+        if (trust && trust->isAuthorized())
+        {
+            return trust->getBalance();
+        }
+        else
+        {
+            return 0;
+        }
+    }
+}
+
+static int64_t
+getAvailableLimit(AccountID const& accountID, Asset const& asset,
+                  int64_t balanceAboveReserve, LedgerDelta& ld, Database& db)
+{
+    if (asset.type() == ASSET_TYPE_NATIVE)
+    {
+        return INT64_MAX - balanceAboveReserve;
+    }
+
+    if (accountID == getIssuer(asset))
+    {
+        return INT64_MAX;
+    }
+    else
+    {
+        auto trust = TrustFrame::loadTrustLine(accountID, asset, db, &ld);
+        if (trust && trust->isAuthorized())
+        {
+            return trust->getTrustLine().limit - trust->getBalance();
+        }
+        else
+        {
+            return 0;
+        }
+    }
+}
+
+static bool
+shouldDeleteOffer(Asset const& asset, int64_t balanceAboveReserve,
+                  std::map<Asset, std::unique_ptr<int64_t>> const& liabilities,
+                  std::function<int64_t(Asset const&, int64_t)> getCap)
+{
+    auto iter = liabilities.find(asset);
+    if (iter == liabilities.end())
+    {
+        throw std::runtime_error("liabilities were not calculated");
+    }
+    // Offers should be deleted if liabilities exceed INT64_MAX (nullptr) or if
+    // there are excess liabilities.
+    return iter->second ? *iter->second > getCap(asset, balanceAboveReserve)
+                        : true;
+}
+
+// This function is used to bring offers and liabilities into a valid state.
+// For every account that has offers,
+//   1. Calculate total liabilities for each asset
+//   2. For every asset with excess buying liabilities according to (1), erase
+//      all offers buying that asset. For every asset with excess selling
+//      liabilities according to (1), erase all offers selling that asset.
+//   3. Update liabilities to reflect offers remaining in the book.
+// It is essential to note that the excess liabilities are determined only
+// using the initial result of step (1), so it does not matter what order the
+// offers are processed.
+static void
+prepareLiabilities(LedgerManager& ledgerManager, LedgerDelta& ld)
```

Comment:
Shouldn't this function also remove offers from accounts with `trustline.isAuthorized() == false`? `AllowTrust` operation removes them when revoking trust so it seems that leaving them is an invalid state.